### ISSUE TITLE
calico/node is not ready: BIRD is not ready: Error querying BIRD: unable to connect to BIRDv4 socket: dial unix /var/run/bird/bird.ctl: connect: no such file or directory

### ISSUE THREAD

**Issue by wangzhao123456** [projectcalico/calico#2834]
I was building a three nodes kubernetes using Calico. But I run into some problem with Calico。
I install Calico-v3.6 according to the instructions in the official website of Calico. Before I join other nodes to the cluster, the Calico pod is running well. But when I join other nodes into the cluster, the Calico pod is not ready but still running. I use "kubectl get nodes" to acquire the state of all nodes and all the nodes is ready. But the Calico pod is unnormal. The pod between nodes can not communicate to each other.
I downloaded the yaml file and the required image of Calico-v3.6, and install Calico using the following command: 
kubecctl apply -f calico.yaml
The version of the kubernetes is v 1.15.2

Thanks in advance!

## Expected Behavior
All the kube-system pod is ready and running. And all the pods in the cluster can communicate to each other.

## Current Behavior
The outcome of "kubectl get pod -n kube-system" are shown as follows:
![image](https://user-images.githubusercontent.com/15125451/64219824-ab710480-cef9-11e9-96b7-f42017689d57.png)
I use "kubectl describe pod calico-node-wh7dr -n kube-system", I got the following message:
![image](https://user-images.githubusercontent.com/15125451/64219912-0acf1480-cefa-11e9-9a5a-8cd8c1e7d0e4.png)

## Steps to Reproduce (for bugs)

All the required condition of k8s is done well.
1. execute "$ kubeadm reset" on all the cluster node.
2. execute on master node: "$ kubeadm init   --kubernetes-version=v1.15.2   --pod-network-cidr=192.168.0.0/16   --apiserver-advertise-address=10.168.104.98 --token-ttl 0"
2.    configure kubectl as a regular user on master node.
       $ mkdir -p $HOME/.kube
       $ sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
       $ sudo chown $(id -u):$(id -g) $HOME/.kube/config
3.    Install pod network Calico on master nodes.
      $ kubectl apply -f calico.yaml
4.    Join into the cluster, execute on other nodes.
      $ kubeadm join 10.168.104.98:6443 --token eh6o8d.0j0xwd0ddty8eakr \
    --discovery-token-ca-cert-hash sha256:659cbac9c7e459cbbf3f42fc4d884a4dc2d5ad1e668750406c04523a7803bcb5 

## Context

Before I join other nodes into the cluster, all the kube-system pods are running well. When I execute join command on other nodes, some calico nodes is not ready but still running, as is shown on above image. The pods on different nodes can not communicate to each other.

## Your Environment

* Calico version: v3.6.5
* k8s version: v1.15.2
* Operating System and version: Ubuntu16.04

The overall error logs of "kubectl describe pod calico-node-wh7dr -n kube-system" command are 
shown as follows:

[wangz@master calico-v3.6]$ kubectl describe pod calico-node-wh7dr -n kube-system
Name:           calico-node-wh7dr
Namespace:      kube-system
Priority:       0
Node:           user08/10.168.103.200
Start Time:     Wed, 04 Sep 2019 09:02:03 +0800
Labels:         controller-revision-hash=66f967c555
                k8s-app=calico-node
                pod-template-generation=1
Annotations:    scheduler.alpha.kubernetes.io/critical-pod: 
Status:         Running
IP:             10.168.103.200
Controlled By:  DaemonSet/calico-node
Init Containers:
  upgrade-ipam:
    Container ID:  docker://27bac175a7cf7e0771e88585cba6a40c8478af58341d759d085fe9686f257aa8
    Image:         calico/cni:v3.6.5
    Image ID:      docker://sha256:0e5467f673396014301fb8a17cfc3aed276ff71ccc595a69950cbcd1e13e3c1f
    Port:          <none>
    Host Port:     <none>
    Command:
      /opt/cni/bin/calico-ipam
      -upgrade
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 04 Sep 2019 09:02:05 +0800
      Finished:     Wed, 04 Sep 2019 09:02:05 +0800
    Ready:          True
    Restart Count:  0
    Environment:
      KUBERNETES_NODE_NAME:        (v1:spec.nodeName)
      CALICO_NETWORKING_BACKEND:  <set to the key 'calico_backend' of config map 'calico-config'>  Optional: false
    Mounts:
      /host/opt/cni/bin from cni-bin-dir (rw)
      /var/lib/cni/networks from host-local-net-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from calico-node-token-nkf26 (ro)
  install-cni:
    Container ID:  docker://b7359636fe9ff7945dccb21214f0b26330504dbe9cbf1b80bdf781f7ea4628cd
    Image:         calico/cni:v3.6.5
    Image ID:      docker://sha256:0e5467f673396014301fb8a17cfc3aed276ff71ccc595a69950cbcd1e13e3c1f
    Port:          <none>
    Host Port:     <none>
    Command:
      /install-cni.sh
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 04 Sep 2019 09:02:06 +0800
      Finished:     Wed, 04 Sep 2019 09:02:06 +0800
    Ready:          True
    Restart Count:  0
    Environment:
      CNI_CONF_NAME:         10-calico.conflist
      CNI_NETWORK_CONFIG:    <set to the key 'cni_network_config' of config map 'calico-config'>  Optional: false
      KUBERNETES_NODE_NAME:   (v1:spec.nodeName)
      CNI_MTU:               <set to the key 'veth_mtu' of config map 'calico-config'>  Optional: false
      SLEEP:                 false
    Mounts:
      /host/etc/cni/net.d from cni-net-dir (rw)
      /host/opt/cni/bin from cni-bin-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from calico-node-token-nkf26 (ro)
Containers:
  calico-node:
    Container ID:   docker://6d4a48784654892a3a7b2672db6d93325ef52b8351b1adc00e1e44dd63c44513
    Image:          calico/node:v3.6.5
    Image ID:       docker://sha256:9b23e1a2ef6d2f5f08903fbc0a18fae61bdb601f39bf8027fe3f94f04c28f88c
    Port:           <none>
    Host Port:      <none>
    State:          Running
      Started:      Wed, 04 Sep 2019 09:02:07 +0800
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:      250m
    Liveness:   http-get http://localhost:9099/liveness delay=10s timeout=1s period=10s #success=1 #failure=6
    Readiness:  exec [/bin/calico-node -bird-ready -felix-ready] delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:
      DATASTORE_TYPE:                     kubernetes
      WAIT_FOR_DATASTORE:                 true
      NODENAME:                            (v1:spec.nodeName)
      CALICO_NETWORKING_BACKEND:          <set to the key 'calico_backend' of config map 'calico-config'>  Optional: false
      CLUSTER_TYPE:                       k8s,bgp
      IP:                                 autodetect
      CALICO_IPV4POOL_IPIP:               Always
      FELIX_IPINIPMTU:                    <set to the key 'veth_mtu' of config map 'calico-config'>  Optional: false
      CALICO_IPV4POOL_CIDR:               192.168.0.0/16
      CALICO_DISABLE_FILE_LOGGING:        true
      FELIX_DEFAULTENDPOINTTOHOSTACTION:  ACCEPT
      FELIX_IPV6SUPPORT:                  false
      FELIX_LOGSEVERITYSCREEN:            info
      FELIX_HEALTHENABLED:                true
    Mounts:
      /lib/modules from lib-modules (ro)
      /run/xtables.lock from xtables-lock (rw)
      /var/lib/calico from var-lib-calico (rw)
      /var/run/calico from var-run-calico (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from calico-node-token-nkf26 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  lib-modules:
    Type:          HostPath (bare host directory volume)
    Path:          /lib/modules
    HostPathType:  
  var-run-calico:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/calico
    HostPathType:  
  var-lib-calico:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/calico
    HostPathType:  
  xtables-lock:
    Type:          HostPath (bare host directory volume)
    Path:          /run/xtables.lock
    HostPathType:  FileOrCreate
  cni-bin-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /opt/cni/bin
    HostPathType:  
  cni-net-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/cni/net.d
    HostPathType:  
  host-local-net-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/cni/networks
    HostPathType:  
  calico-node-token-nkf26:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  calico-node-token-nkf26
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  beta.kubernetes.io/os=linux
Tolerations:     :NoSchedule
                 :NoExecute
                 CriticalAddonsOnly
                 node.kubernetes.io/disk-pressure:NoSchedule
                 node.kubernetes.io/memory-pressure:NoSchedule
                 node.kubernetes.io/network-unavailable:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute
                 node.kubernetes.io/pid-pressure:NoSchedule
                 node.kubernetes.io/unreachable:NoExecute
                 node.kubernetes.io/unschedulable:NoSchedule
Events:
  Type     Reason     Age   From             Message
  ----     ------     ----  ----             -------
  Normal   Pulled     51m   kubelet, user08  Container image "calico/cni:v3.6.5" already present on machine
  Normal   Created    51m   kubelet, user08  Created container upgrade-ipam
  Normal   Started    51m   kubelet, user08  Started container upgrade-ipam
  Normal   Pulled     51m   kubelet, user08  Container image "calico/cni:v3.6.5" already present on machine
  Normal   Created    51m   kubelet, user08  Created container install-cni
  Normal   Started    51m   kubelet, user08  Started container install-cni
  Normal   Pulled     51m   kubelet, user08  Container image "calico/node:v3.6.5" already present on machine
  Normal   Created    51m   kubelet, user08  Created container calico-node
  Normal   Started    51m   kubelet, user08  Started container calico-node
  Warning  Unhealthy  51m   kubelet, user08  Readiness probe failed: Threshold time for bird readiness check:  30s
calico/node is not ready: BIRD is not ready: Error querying BIRD: unable to connect to BIRDv4 socket: dial unix /var/run/bird/bird.ctl: connect: no such file or directory
  Normal   Scheduled  51m  default-scheduler  Successfully assigned kube-system/calico-node-wh7dr to user08
  Warning  Unhealthy  51m  kubelet, user08    Readiness probe failed: Threshold time for bird readiness check:  30s
calico/node is not ready: BIRD is not ready: BGP not established with 192.168.0.2422019-09-04 01:02:19.322 [INFO][278] readiness.go 88: Number of node(s) with BGP peering established = 0
  Warning  Unhealthy  51m  kubelet, user08  Readiness probe failed: Threshold time for bird readiness check:  30s
calico/node is not ready: BIRD is not ready: BGP not established with 192.168.0.2422019-09-04 01:02:29.318 [INFO][315] readiness.go 88: Number of node(s) with BGP peering established = 0
  Warning  Unhealthy  51m  kubelet, user08  Readiness probe failed: Threshold time for bird readiness check:  30s
calico/node is not ready: BIRD is not ready: BGP not established with 192.168.0.2422019-09-04 01:02:39.330 [INFO][341] readiness.go 88: Number of node(s) with BGP peering established = 0
  Warning  Unhealthy  50m  kubelet, user08  Readiness probe failed: Threshold time for bird readiness check:  30s
calico/node is not ready: BIRD is not ready: BGP not established with 192.168.0.2422019-09-04 01:02:49.268 [INFO][377] readiness.go 88: Number of node(s) with BGP peering established = 0
  Warning  Unhealthy  50m  kubelet, user08  Readiness probe failed: Threshold time for bird readiness check:  30s
calico/node is not ready: BIRD is not ready: BGP not established with 192.168.0.2422019-09-04 01:02:59.247 [INFO][407] readiness.go 88: Number of node(s) with BGP peering established = 0
  Warning  Unhealthy  50m  kubelet, user08  Readiness probe failed: Threshold time for bird readiness check:  30s
calico/node is not ready: BIRD is not ready: BGP not established with 192.168.0.2422019-09-04 01:03:09.264 [INFO][430] readiness.go 88: Number of node(s) with BGP peering established = 0
  Warning  Unhealthy  50m  kubelet, user08  Readiness probe failed: Threshold time for bird readiness check:  30s
calico/node is not ready: BIRD is not ready: BGP not established with 192.168.0.2422019-09-04 01:03:19.256 [INFO][462] readiness.go 88: Number of node(s) with BGP peering established = 0
  Warning  Unhealthy  50m  kubelet, user08  Readiness probe failed: Threshold time for bird readiness check:  30s
calico/node is not ready: BIRD is not ready: BGP not established with 192.168.0.2422019-09-04 01:03:29.244 [INFO][489] readiness.go 88: Number of node(s) with BGP peering established = 0
  Warning  Unhealthy  91s (x292 over 50m)  kubelet, user08  (combined from similar events): Readiness probe failed: Threshold time for bird readiness check:  30s
calico/node is not ready: BIRD is not ready: BGP not established with 192.168.0.2422019-09-04 01:52:09.237 [INFO][8711] readiness.go 88: Number of node(s) with BGP peering established = 0

**Comment by tmjd** [comment]
Do you hosts have multiple IP addresses? I'm guessing yes and that calico-node is selecting the wrong IP address to use. This is based on the BGP not established being for 192.168.0.2 yet the describe pod for calico-node (which is host networked) has 10.168.103.200 and also I see that your apiserver address is 10.168.103.200. So I'm guessing your hosts should be using addresses in 10.168.x.x but at least one is using 192.168.0.2.  You should take a look at https://docs.projectcalico.org/v3.8/reference/node/configuration#ip-setting, specifically IP autodetection as I expect you need to configure that.

**Comment by JLoveI** [comment]
I have the same problem, any solutions?

**Comment by dstdfx** [comment]
@JLoveI worked for me: https://github.com/projectcalico/calico/issues/2042#issuecomment-408488357

**Comment by shaopenghui** [comment]
I also have the same problem
The following logs:[ERROR][59] health.go 238: Health endpoint failed, trying to restart it... error=listen tcp 127.0.0.1:9099: bind: address already in use
kubectl describe pod:Readiness probe failed: calico/node is not ready: BIRD is not ready: Error querying BIRD: unable to connect to BIRDv4 socket: dial unix /var/run/bird/bird.ctl: connect: no such file or directory

**Comment by caseydavenport** [comment]
>  calico/node is not ready: BIRD is not ready: Error querying BIRD: unable to connect to BIRDv4 socket: dial unix /var/run/bird/bird.ctl: connect: no such file or directory

This appeared to be a transient issue in the original post. Agree with @tmjd that this feels like a misconfigured IP address auto-detection issue (and is supported by this comment as well: https://github.com/projectcalico/calico/issues/2834#issuecomment-542645416) 

Also worth noting that we improved BIRD liveness reporting in our most recent release which might help in this area.

**Comment by GustavGenberg** [comment]
I had the same problem with "calico/node is not ready: BIRD is not ready: Error querying BIRD: unable to connect to BIRDv4 socket: dial unix /var/run/bird/bird.ctl: connect: no such file or directory".

I checked the logs and the calico node was complaining (not an error, just a warning) that the kernel version was too old (running Ubuntu Server 18.04, kernel version 4.15.0-96-generic, it wanted 4.16 I think) and something failed. I did a distro upgrade to Ubuntu version 20.04 (kernel version 5.4.0-58-generic), and now it just works.
